src/parse/lex.rs

```diff
@@ -1,22 +1,319 @@
+
+    }
+
+    ///double bump
+    fn dbump(&mut self) {
+        self.bump();
+        self.bump();
+    }
+
+    ///scan each new WORD from the query string
+    fn scan_words(&mut self) -> String {
+        let mut s = String::new();
+        //loop until the end of a word (only letters and _)
+        loop {
+            //take current char
+            match self.curr.unwrap_or(' ') {
+                c @ 'a' ... 'z' |
+                c @ 'A' ... 'Z' |
+                c @ '0' ... '9' |
+                c @ '_' => {
+                    //push letter into return string
+                    s.push(c);
+                },
+                //if no letter, then stop loop
+                _ => break
+            }
+            //next char
+            self.bump();
+        }
+        s
+    }
+
+    ///scan each new NUMBER from the query string
+    fn scan_nums(&mut self) -> String {
+        let mut s = String::new();
+        let mut dot = false;
+        loop {
+            match self.curr.unwrap_or(' ') {
+                c @ '0' ... '9' => {
+                    s.push(c);
+                }
+                c @ '.' => {
+                    if dot {
+                        println!("This is not a good number, Sir!");
+                    }
+                    dot = true;
+                    s.push(c);
+                },
+                _ => break
+            }
+            self.bump();
         }
+        s
     }
 
-    pub fn next(&mut self) -> Option<TokenSpan> {
-        None
+    fn scan_lit(&mut self) -> String {
+        //liter = everything until " | ' , bump over ' | "
+        let mut s = String::new();
+        self.bump(); //to first char of literal
+        loop {
+            match self.curr.unwrap_or(' ') {
+                c @ '\'' |
+                c @ '"' => {
+                    break
+                }
+                c @ _ => {
+                    s.push(c);
+                }
+            }
+            self.bump();
+        }
+        s
+    }
+
+    ///skips all the whitespaces
+    fn skip_whitespace(&mut self) {
+        while is_whitespace(self.curr.unwrap()){
+            self.bump();
+        }
     }
-}
\ No newline at end of file
+}
+
+///checks for whitespace/line break/tab
+fn is_whitespace(c: char) -> bool {
+    match c {
+        ' ' | '\n' | '\t' => true,
+        _ => false
+    }
+}
+
+impl<'a> Iterator for Lexer<'a> {
+
+    type Item = TokenSpan;
+
+    fn next(&mut self) -> Option<TokenSpan> {
+
+        //for two char ops (e.g. >=), we need to check the next char
+        //if we are at the end of the string, we end
+        let nexchar = self.next.unwrap_or('\x00');
+
+        //saving the first pos of a token
+        self.span_start = self.curr_pos;
+
+        //if no current char available, return None
+        if self.curr.is_none() {
+            return None;
+        }
+
+        //unwrap current char and decide token class
+        let token = match self.curr.unwrap_or('x') {
```

Comment:
Instead of this strange `unwrap_or`, you should change the if block above to:
```rust
let curr = match self.curr {
    None => return None,
    Some(c) => c,
};
```
I hope it's clear how this works and why this is better.